# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2005/01/04 15:57:25-08:00 brugolsky@telemetry-investments.com 
#   [PATCH] NFS client O_DIRECT error case fix
#   
#   The NFS direct-io error return path for request sizes greater than
#   MAX_DIRECTIO_SIZE fails to initialize the returned page struct array
#   pointer to NULL.
#   
#   Discovered using AKPM's ext3-tools: odwrite -ko 0 16385 foo
#   
#   Signed-off-by: Bill Rugolsky <brugolsky@telemetry-investments.com>
#   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
# 
# fs/nfs/direct.c
#   2005/01/04 13:55:37-08:00 brugolsky@telemetry-investments.com +3 -1
#   NFS client O_DIRECT error case fix
# 
diff -Nru a/fs/nfs/direct.c b/fs/nfs/direct.c
--- a/fs/nfs/direct.c	2005-01-04 20:03:28 -08:00
+++ b/fs/nfs/direct.c	2005-01-04 20:03:28 -08:00
@@ -72,8 +72,10 @@
 	size_t array_size;
 
 	/* set an arbitrary limit to prevent arithmetic overflow */
-	if (size > MAX_DIRECTIO_SIZE)
+	if (size > MAX_DIRECTIO_SIZE) {
+		*pages = NULL;
 		return -EFBIG;
+	}
 
 	page_count = (user_addr + size + PAGE_SIZE - 1) >> PAGE_SHIFT;
 	page_count -= user_addr >> PAGE_SHIFT;
